File: apps/web/src/store/dashboard-store.ts
'use client';

import { create } from 'zustand';
import { statisticsService, budgetService, transactionService } from '@/lib/api-services';
import { formatDate } from '@/lib/utils';
import { apiClient } from '@/lib/api-client';
import dayjs from 'dayjs';

// 仪表盘状态类型
interface DashboardState {
  // 数据状态
  monthlyStats: {
    income: number;
    expense: number;
    balance: number;
    month: string;
  };
  budgetCategories: any[];
  totalBudget: any;
  groupedTransactions: any[];
  isLoading: boolean;
  error: string | null;

  // 分页加载相关状态
  isLoadingMore: boolean;
  hasMoreTransactions: boolean;
  currentPage: number;
  totalTransactionsCount: number;
  autoRefreshCount: number; // 自动刷新计数
  showBackToTop: boolean; // 是否显示返回顶部按钮

  // 操作方法
  fetchDashboardData: (accountBookId: string) => Promise<void>;
  refreshDashboardData: (accountBookId: string) => Promise<void>;
  clearDashboardData: () => void;
  // 新增：监听记账变化的方法
  setupTransactionListener: () => void;
  cleanupTransactionListener: () => void;

  // 新增：分页加载方法
  loadMoreTransactions: (accountBookId: string) => Promise<void>;
  resetTransactionPagination: () => void;
  incrementAutoRefreshCount: () => void;
  setShowBackToTop: (show: boolean) => void;
}

// 获取月度统计的辅助函数
const fetchMonthlyStatistics = async (accountBookId: string) => {
  console.log('开始获取月度统计数据...');
  const startDate = dayjs().startOf('month').format('YYYY-MM-DD');
  const endDate = dayjs().endOf('month').format('YYYY-MM-DD');

  const response = await statisticsService.getStatistics(accountBookId, {
    startDate,
    endDate,
  });

  console.log('月度统计数据响应:', response);

  return {
    income: response?.income || 0,
    expense: response?.expense || 0,
    balance: response?.netIncome || 0,
    month: formatDate(new Date(), 'YYYY年MM月'),
  };
};

// 获取预算统计的辅助函数
const fetchBudgetStatistics = async (accountBookId: string) => {
  console.log('开始获取预算统计数据...');
  const currentMonth = dayjs().format('YYYY-MM');

  const response = await budgetService.getBudgetStatistics(accountBookId, {
    month: currentMonth,
  });

  console.log('预算统计数据响应:', response);

  const categories =
    response?.categories?.map((cat: any) => ({
      id: cat.category.id,
      name: cat.category.name,
      icon: cat.category.icon,
      budget: cat.budget,
      spent: cat.spent,
      percentage: cat.percentage,
      period: cat.period || 'MONTHLY',
      categoryId: cat.category.id,
    })) || [];

  // 直接使用后端返回的百分比，不重新计算
  const totalBudget =
    response?.totalBudget && response?.totalSpent !== undefined
      ? {
          amount: response.totalBudget,
          spent: response.totalSpent,
          percentage: response.percentage || 0, // 直接使用后端计算的百分比
        }
      : null;

  console.log('处理后的总预算数据:', totalBudget);

  return { categories, totalBudget };
};

// 获取最近记账的辅助函数
const fetchRecentTransactions = async (accountBookId: string, page: number = 1, limit: number = 20) => {
  console.log(`开始获取最近记账数据，页码: ${page}, 每页: ${limit}...`);

  // 使用通用的transactions接口，支持分页
  const transactionsResponse = await apiClient.get('/transactions', {
    params: {
      accountBookId,
      page,
      limit,
      sort: 'date:desc',
      includeAttachments: true,
    },
  });
  console.log('最近记账数据响应:', transactionsResponse);

  if (transactionsResponse?.data && Array.isArray(transactionsResponse.data)) {
    const groupedByDate: Record<string, any[]> = {};

    transactionsResponse.data.forEach((tx: any) => {
      const dateKey = dayjs(tx.date).format('YYYY-MM-DD');
      if (!groupedByDate[dateKey]) {
        groupedByDate[dateKey] = [];
      }
      groupedByDate[dateKey].push(tx);
    });

    const formattedTransactions = Object.keys(groupedByDate)
      .sort((a, b) => new Date(b).getTime() - new Date(a).getTime())
      .map((dateKey) => {
        return {
          date: formatDate(dateKey, 'MM月DD日'),
          transactions: groupedByDate[dateKey].map((tx: any) => ({
            id: tx.id,
            amount: tx.amount,
            type: tx.type,
            categoryName: tx.category?.name || '未分类',
            categoryIcon: tx.category?.icon || 'other',
            description: tx.description || '',
            date: tx.date,
            category: tx.category,
            tags: tx.tags,
            attachments: tx.attachments || [], // 保留附件信息
            attachmentCount: tx.attachmentCount || 0, // 保留附件数量
          })),
        };
      });

    console.log('格式化后的记账数据:', formattedTransactions);

    // 返回格式化后的数据和分页信息
    return {
      transactions: formattedTransactions,
      hasMore: transactionsResponse.data.length === limit &&
                page * limit < (transactionsResponse.total || 0),
      total: transactionsResponse.total || 0,
      currentPage: page,
    };
  }

  // 返回空数据和分页信息
  return {
    transactions: [],
    hasMore: false,
    total: 0,
    currentPage: page,
  };
};

// 创建仪表盘状态管理
export const useDashboardStore = create<DashboardState>((set, get) => {
  // 事件处理函数，需要在外部定义以便正确移除
  let transactionChangeHandler: ((event: CustomEvent) => void) | null = null;

  return {
    // 初始状态
    monthlyStats: {
      income: 0,
      expense: 0,
      balance: 0,
      month: formatDate(new Date(), 'YYYY年MM月'),
    },
    budgetCategories: [],
    totalBudget: null,
    groupedTransactions: [],
    isLoading: false,
    error: null,

    // 分页加载相关状态
    isLoadingMore: false,
    hasMoreTransactions: true,
    currentPage: 1,
    totalTransactionsCount: 0,
    autoRefreshCount: 0,
    showBackToTop: false,

    // 获取仪表盘数据
    fetchDashboardData: async (accountBookId: string) => {
      try {
        set({ isLoading: true, error: null, currentPage: 1 });

        // 并行请求数据
        const [monthlyStats, budgetData, transactionData] = await Promise.all([
          fetchMonthlyStatistics(accountBookId),
          fetchBudgetStatistics(accountBookId),
          fetchRecentTransactions(accountBookId, 1, 20),
        ]);

        set({
          monthlyStats,
          budgetCategories: budgetData.categories,
          totalBudget: budgetData.totalBudget,
          groupedTransactions: transactionData.transactions,
          hasMoreTransactions: transactionData.hasMore,
          totalTransactionsCount: transactionData.total,
          currentPage: transactionData.currentPage,
          isLoading: false,
        });
      } catch (error) {
        console.error('获取仪表盘数据失败:', error);
        set({
          isLoading: false,
          error: '获取仪表盘数据失败',
        });
      }
    },

    // 刷新仪表盘数据（不显示加载状态）
    refreshDashboardData: async (accountBookId: string) => {
      try {
        set({ error: null, currentPage: 1 });
        console.log(`开始刷新仪表盘数据，账本ID: ${accountBookId}`);

        // 注意：当前使用的apiClient不支持缓存，所以总是获取最新数据
        console.log('开始获取最新数据...');

        // 并行请求数据
        const [monthlyStats, budgetData, transactionData] = await Promise.all([
          fetchMonthlyStatistics(accountBookId),
          fetchBudgetStatistics(accountBookId),
          fetchRecentTransactions(accountBookId, 1, 20),
        ]);

        console.log('所有数据获取完成，更新状态...');
        set({
          monthlyStats,
          budgetCategories: budgetData.categories,
          totalBudget: budgetData.totalBudget,
          groupedTransactions: transactionData.transactions,
          hasMoreTransactions: transactionData.hasMore,
          totalTransactionsCount: transactionData.total,
          currentPage: transactionData.currentPage,
        });
        console.log('仪表盘状态更新完成');
      } catch (error) {
        console.error('刷新仪表盘数据失败:', error);
        set({ error: '刷新仪表盘数据失败' });
        // 重新抛出异常，让调用者知道刷新失败
        throw error;
      }
    },

    // 清空仪表盘数据
    clearDashboardData: () => {
      set({
        monthlyStats: {
          income: 0,
          expense: 0,
          balance: 0,
          month: formatDate(new Date(), 'YYYY年MM月'),
        },
        budgetCategories: [],
        totalBudget: null,
        groupedTransactions: [],
        isLoading: false,
        error: null,

        // 重置分页状态
        isLoadingMore: false,
        hasMoreTransactions: true,
        currentPage: 1,
        totalTransactionsCount: 0,
        autoRefreshCount: 0,
        showBackToTop: false,
      });
    },

    // 加载更多交易记录
    loadMoreTransactions: async (accountBookId: string) => {
      const currentState = get();

      // 如果正在加载或没有更多数据，直接返回
      if (currentState.isLoadingMore || !currentState.hasMoreTransactions || currentState.isLoading) {
        console.log('⏸️ [Dashboard Store] 跳过加载更多：', {
          isLoadingMore: currentState.isLoadingMore,
          hasMoreTransactions: currentState.hasMoreTransactions,
          isLoading: currentState.isLoading
        });
        return;
      }

      try {
        console.log(`🔄 [Dashboard Store] 加载更多交易记录，当前页: ${currentState.currentPage}, 当前加载次数: ${currentState.autoRefreshCount}`);

        // 立即设置加载状态，防止重复触发
        set({ isLoadingMore: true });

        const nextPage = currentState.currentPage + 1;
        const transactionData = await fetchRecentTransactions(accountBookId, nextPage, 20);

        // 合并新旧交易记录
        const updatedTransactions = [...currentState.groupedTransactions, ...transactionData.transactions];

        // 增加加载计数
        const newAutoRefreshCount = currentState.autoRefreshCount + 1;

        set({
          groupedTransactions: updatedTransactions,
          hasMoreTransactions: transactionData.hasMore,
          totalTransactionsCount: transactionData.total,
          currentPage: transactionData.currentPage,
          isLoadingMore: false,
          autoRefreshCount: newAutoRefreshCount, // 更新加载计数
        });

        console.log(`✅ [Dashboard Store] 成功加载更多交易记录，新记录数: ${transactionData.transactions.length}，还有更多: ${transactionData.hasMore}，加载次数: ${newAutoRefreshCount}`);
      } catch (error) {
        console.error('❌ [Dashboard Store] 加载更多交易记录失败:', error);
        set({ isLoadingMore: false });
      }
    },

    // 重置交易分页状态
    resetTransactionPagination: () => {
      set({
        isLoadingMore: false,
        hasMoreTransactions: true,
        currentPage: 1,
        totalTransactionsCount: 0,
        autoRefreshCount: 0,
        showBackToTop: false,
      });
    },

    // 增加自动刷新计数
    incrementAutoRefreshCount: () => {
      const currentState = get();
      const newCount = currentState.autoRefreshCount + 1;
      set({ autoRefreshCount: newCount });
    },

    // 设置是否显示返回顶部按钮
    setShowBackToTop: (show: boolean) => {
      set({ showBackToTop: show });
    },

    // 设置记账变化监听器
    setupTransactionListener: () => {
      // 确保在浏览器环境中
      if (typeof window === 'undefined') {
        console.warn('setupTransactionListener: 不在浏览器环境中，跳过监听器设置');
        return;
      }

      // 如果已经有监听器，先清理
      if (transactionChangeHandler) {
        window.removeEventListener('transactionChanged', transactionChangeHandler as EventListener);
        console.log('清理旧的记账变化监听器');
      }

      // 创建新的事件处理函数
      transactionChangeHandler = (event: CustomEvent) => {
        const { accountBookId } = event.detail;
        console.log('监听到记账变化事件，账本ID:', accountBookId);

        // 延迟刷新，确保数据库操作已完成
        setTimeout(() => {
          console.log('开始自动刷新仪表盘数据...');
          get()
            .refreshDashboardData(accountBookId)
            .catch((error) => {
              console.error('自动刷新仪表盘数据失败:', error);
            });
        }, 500);
      };

      window.addEventListener('transactionChanged', transactionChangeHandler as EventListener);
      console.log('仪表盘记账变化监听器已设置');
    },

    // 清理记账变化监听器
    cleanupTransactionListener: () => {
      // 确保在浏览器环境中
      if (typeof window === 'undefined') {
        return;
      }

      if (transactionChangeHandler) {
        window.removeEventListener('transactionChanged', transactionChangeHandler as EventListener);
        transactionChangeHandler = null;
        console.log('仪表盘记账变化监听器已清理');
      }
    },
  };
});

// 全局函数：触发记账变化事件
export const triggerTransactionChange = (accountBookId: string) => {
  // 确保在浏览器环境中
  if (typeof window === 'undefined') {
    console.warn('triggerTransactionChange: 不在浏览器环境中，跳过事件触发');
    return;
  }

  console.log('触发记账变化事件，账本ID:', accountBookId);

  // 方法1：使用自定义事件
  const event = new CustomEvent('transactionChanged', {
    detail: { accountBookId },
  });
  window.dispatchEvent(event);
  console.log('记账变化事件已触发');

  // 方法2：使用localStorage作为备用机制
  const refreshSignal = {
    accountBookId,
    timestamp: Date.now(),
    action: 'refresh_dashboard',
  };
  localStorage.setItem('dashboard_refresh_signal', JSON.stringify(refreshSignal));
  console.log('仪表盘刷新信号已写入localStorage');
};

File: apps/android/app/src/main/java/com/zhiweijizhangandroid/store/DashboardStore.kt
package com.zhiweijizhangandroid.store

import android.content.Context
import com.zhiweijizhangandroid.api.ApiClient
import com.zhiweijizhangandroid.models.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.text.SimpleDateFormat
import java.util.*

/**
 * 仪表盘状态管理
 * 对应Web版本的useDashboardStore
 */
class DashboardStore(
    private val context: Context,
    private val apiClient: ApiClient
) {
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    // 状态流
    private val _dashboardData = MutableStateFlow(DashboardData.empty())
    val dashboardData: StateFlow<DashboardData> = _dashboardData.asStateFlow()

    // 当前账本ID
    private var currentAccountBookId: String? = null

    /**
     * 获取仪表盘数据
     */
    fun fetchDashboardData(accountBookId: String) {
        currentAccountBookId = accountBookId
        println("[DashboardStore] 开始获取仪表盘数据，账本ID: $accountBookId")

        scope.launch {
            try {
                _dashboardData.value = _dashboardData.value.copy(isLoading = true, error = null)

                // 并行请求数据，每个请求独立处理错误
                println("[DashboardStore] 开始并行请求API数据...")

                // 月度统计（必需）
                val monthlyStats = try {
                    fetchMonthlyStatistics(accountBookId)
                } catch (e: Exception) {
                    println("[DashboardStore] 月度统计获取失败，使用默认值: ${e.message}")
                    MonthlyStats(
                        income = 0.0,
                        expense = 0.0,
                        balance = 0.0,
                        month = getCurrentMonth()
                    )
                }

                // 预算数据（可选）
                val budgetData = try {
                    fetchBudgetStatistics(accountBookId)
                } catch (e: Exception) {
                    println("[DashboardStore] 预算数据获取失败，使用空数据: ${e.message}")
                    Pair(emptyList<BudgetCategory>(), null)
                }

                // 交易数据（可选）
                val transactions = try {
                    fetchRecentTransactions(accountBookId)
                } catch (e: Exception) {
                    println("[DashboardStore] 交易数据获取失败，使用空数据: ${e.message}")
                    emptyList<GroupedTransactions>()
                }

                println("[DashboardStore] 所有API请求完成")
                println("[DashboardStore] 月度统计: 收入=${monthlyStats.income}, 支出=${monthlyStats.expense}")
                println("[DashboardStore] 预算分类数量: ${budgetData.first.size}")
                println("[DashboardStore] 交易分组数量: ${transactions.size}")

                _dashboardData.value = DashboardData(
                    monthlyStats = monthlyStats,
                    budgetCategories = budgetData.first,
                    totalBudget = budgetData.second,
                    groupedTransactions = transactions,
                    isLoading = false,
                    error = null
                )

            } catch (e: Exception) {
                println("[DashboardStore] 获取仪表盘数据失败: ${e.message}")
                e.printStackTrace()
                _dashboardData.value = DashboardData.error("获取仪表盘数据失败: ${e.message}")
            }
        }
    }

    /**
     * 刷新仪表盘数据（不显示加载状态）
     */
    fun refreshDashboardData(accountBookId: String) {
        currentAccountBookId = accountBookId

        scope.launch {
            try {
                _dashboardData.value = _dashboardData.value.copy(error = null)

                // 清除缓存
                apiClient.clearCache()

                // 并行请求数据
                val monthlyStatsDeferred = async { fetchMonthlyStatistics(accountBookId) }
                val budgetDataDeferred = async { fetchBudgetStatistics(accountBookId) }
                val transactionsDeferred = async { fetchRecentTransactions(accountBookId) }

                val monthlyStats = monthlyStatsDeferred.await()
                val budgetData = budgetDataDeferred.await()
                val transactions = transactionsDeferred.await()

                _dashboardData.value = _dashboardData.value.copy(
                    monthlyStats = monthlyStats,
                    budgetCategories = budgetData.first,
                    totalBudget = budgetData.second,
                    groupedTransactions = transactions
                )

            } catch (e: Exception) {
                println("刷新仪表盘数据失败: ${e.message}")
                _dashboardData.value = _dashboardData.value.copy(error = "刷新数据失败: ${e.message}")
            }
        }
    }

    /**
     * 清除仪表盘数据
     */
    fun clearDashboardData() {
        _dashboardData.value = DashboardData.empty()
        currentAccountBookId = null
    }

    /**
     * 获取月度统计数据
     */
    private suspend fun fetchMonthlyStatistics(accountBookId: String): MonthlyStats {
        return withContext(Dispatchers.IO) {
            try {
                val calendar = Calendar.getInstance()
                val startDate = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(
                    calendar.apply {
                        set(Calendar.DAY_OF_MONTH, 1)
                    }.time
                )

                calendar.set(Calendar.DAY_OF_MONTH, calendar.getActualMaximum(Calendar.DAY_OF_MONTH))
                val endDate = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(calendar.time)

                println("[DashboardStore] 请求月度统计: $startDate 到 $endDate")
                val response = apiClient.getStatistics(accountBookId, startDate, endDate)
                println("[DashboardStore] 月度统计响应: $response")

                MonthlyStats(
                    income = response?.get("income") as? Double ?: 0.0,
                    expense = response?.get("expense") as? Double ?: 0.0,
                    balance = response?.get("netIncome") as? Double ?: 0.0,
                    month = getCurrentMonth()
                )
            } catch (e: Exception) {
                println("[DashboardStore] 获取月度统计失败: ${e.message}")
                e.printStackTrace()
                throw e
            }
        }
    }

    /**
     * 获取预算统计数据
     */
    private suspend fun fetchBudgetStatistics(accountBookId: String): Pair<List<BudgetCategory>, TotalBudget?> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiClient.getBudgetStatistics(accountBookId)

                val categories = (response?.get("categories") as? List<Map<String, Any>>)?.map { categoryData ->
                    BudgetCategory(
                        id = categoryData["id"] as? String ?: "",
                        name = categoryData["name"] as? String ?: "",
                        icon = categoryData["icon"] as? String,
                        budget = (categoryData["budget"] as? Number)?.toDouble() ?: 0.0,
                        spent = (categoryData["spent"] as? Number)?.toDouble() ?: 0.0,
                        percentage = (categoryData["percentage"] as? Number)?.toDouble() ?: 0.0,
                        period = categoryData["period"] as? String,
                        categoryId = categoryData["categoryId"] as? String
                    )
                } ?: emptyList()

                val totalBudgetData = response?.get("totalBudget") as? Map<String, Any>
                val totalBudget = totalBudgetData?.let {
                    TotalBudget(
                        amount = (it["amount"] as? Number)?.toDouble() ?: 0.0,
                        spent = (it["spent"] as? Number)?.toDouble() ?: 0.0,
                        percentage = (it["percentage"] as? Number)?.toDouble() ?: 0.0
                    )
                }

                Pair(categories, totalBudget)
            } catch (e: Exception) {
                println("[DashboardStore] 获取预算统计失败: ${e.message}")
                // 预算不存在是正常情况，返回空数据而不是抛异常
                if (e.message?.contains("预算不存在") == true || e.message?.contains("404") == true) {
                    println("[DashboardStore] 预算不存在，返回空预算数据")
                    return@withContext Pair(emptyList(), null)
                }
                e.printStackTrace()
                throw e
            }
        }
    }

    /**
     * 获取最近交易数据
     */
    private suspend fun fetchRecentTransactions(accountBookId: String): List<GroupedTransactions> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiClient.getRecentTransactions(accountBookId, limit = 20)
                println("[DashboardStore] 交易API响应: $response")

                // 服务器返回格式: {"total":0,"page":1,"limit":20,"data":[]}
                // 需要从data字段获取交易数据
                val transactionList = response?.get("data") as? List<Map<String, Any>> ?: emptyList()
                println("[DashboardStore] 解析到 ${transactionList.size} 条交易记录")

                val transactions = transactionList.map { txData ->
                    Transaction(
                        id = txData["id"] as? String ?: "",
                        amount = (txData["amount"] as? Number)?.toDouble() ?: 0.0,
                        type = if ((txData["type"] as? String) == "EXPENSE") TransactionType.EXPENSE else TransactionType.INCOME,
                        categoryName = (txData["category"] as? Map<String, Any>)?.get("name") as? String ?: "未分类",
                        categoryIcon = (txData["category"] as? Map<String, Any>)?.get("icon") as? String,
                        description = txData["description"] as? String ?: "",
                        date = txData["date"] as? String ?: ""
                    )
                }

                // 按日期分组
                val grouped = groupTransactionsByDate(transactions)
                println("[DashboardStore] 分组后的交易数据: ${grouped.size} 组")
                grouped
            } catch (e: Exception) {
                println("[DashboardStore] 获取最近交易失败: ${e.message}")
                e.printStackTrace()
                // 交易数据获取失败时返回空列表，不影响其他数据显示
                emptyList()
            }
        }
    }

    /**
     * 按日期分组交易
     */
    private fun groupTransactionsByDate(transactions: List<Transaction>): List<GroupedTransactions> {
        return transactions.groupBy { transaction ->
            // 提取日期部分 (YYYY-MM-DD)，忽略时间部分
            try {
                // 处理ISO 8601格式: 2025-05-31T03:18:44.193Z
                val dateOnly = transaction.date.substringBefore("T")
                println("[DashboardStore] 交易日期: ${transaction.date} -> 分组日期: $dateOnly")
                dateOnly
            } catch (e: Exception) {
                println("[DashboardStore] 日期解析失败: ${transaction.date}, 使用原始值")
                transaction.date
            }
        }.map { (dateKey, txList) ->
            // 格式化显示日期
            val displayDate = try {
                // 将 2025-05-31 格式化为 05月31日
                val parts = dateKey.split("-")
                if (parts.size >= 3) {
                    "${parts[1]}月${parts[2]}日"
                } else {
                    dateKey
                }
            } catch (e: Exception) {
                dateKey
            }

            println("[DashboardStore] 日期分组: $dateKey -> $displayDate (${txList.size}条交易)")
            GroupedTransactions(date = displayDate, transactions = txList.sortedByDescending { it.date })
        }.sortedByDescending { group ->
            // 按原始日期排序
            try {
                group.transactions.firstOrNull()?.date ?: ""
            } catch (e: Exception) {
                group.date
            }
        }
    }

    /**
     * 获取当前月份字符串
     */
    private fun getCurrentMonth(): String {
        val calendar = Calendar.getInstance()
        val year = calendar.get(Calendar.YEAR)
        val month = calendar.get(Calendar.MONTH) + 1
        return "${year}年${String.format("%02d", month)}月"
    }

    /**
     * 清理资源
     */
    fun cleanup() {
        scope.cancel()
    }
}
